<?xml version="1.0" ?>
<launch>
	<!-- If mode = sim: Simulated AUV with MBES and control in loaded bathymetric map 
	If mode = gt: reproduce Hugin survey and construct map of real bathymetry --> 
  <arg name="mode" default="gt"/>
	<arg name="namespace" default="hugin_0"/>
	<arg name="debug" default="0"/>
	<arg name="dataset" default="ripples"/>
	<arg name="app" default="rbpf"/>
  <!-- <arg name="path" default="$(find uw_tests)/datasets/$(arg dataset)/"/> -->

  <!-- Num of beams per MBES pings to store in the log -->
	<arg name="n_beams_mbes" default="100"/>	

  <!-- Coordinates of odom frame wrt the map frame. By default, the AUV start the mission here -->
	<!-- <arg name="x" default="-300.0"/>
	<arg name="y" default="-400.0"/>
	<arg name="z" default="0.0"/>
	<arg name="roll" default="0.0"/>
	<arg name="pitch" default="0.0"/>
	<arg name="yaw" default="0.0"/> -->

  <!-- RBPF params  -->
  <arg name="particle_count"                    default="10"/> <!--50 -->
  <arg name="particle_count_neighbours" default="5"/>
  <arg name="num_particle_handlers"             default="1"/> 
  <arg name="survey_finished_top" default="/$(arg mode)/survey_finished"/>
  
  <!-- API for SVGP particles -->
  <arg name="gp_ip_topic" default="/$(arg namespace)/gp/inducing_points"/>
  <arg name="gp_mb_server" default="/$(arg namespace)/gp/minibatch_server"/>
  <!-- <arg name="gp_plot_server" default="/gp/plot_server"/> -->
  <arg name="manipulate_gp_server" default="/$(arg namespace)/gp/manipulate_server"/>
  <arg name="gp_resample_top" default="/$(arg namespace)/gp/resample_top"/>
  
  <arg name="results_path" default=""/>     
  <arg name="rbpf_period"   default="15." />
  <arg name="rviz_period"   default="1." />
  <arg name="svgp_minibatch_size"   default="1000" />

  <arg name="map_frame"                         default="map"/>
  <arg name="odom_frame"                        default="$(arg namespace)/odom"/>
  <arg name="odometry_topic"                    default="/$(arg mode)/$(arg namespace)/odom"/>
  <!-- <arg name="odometry_topic"                    default="/lolo/dr/odom_fixed"/> -->
  <arg name="mbes_pings_topic"                  default="/$(arg mode)/$(arg namespace)/mbes_pings"/>
  <!-- <arg name="mbes_pings_topic"                  default="/rbpf/mbes_pings"/> -->
  <arg name="average_pose_topic"                default="/$(arg namespace)/$(arg app)/avg_pose"/>
  <arg name="particle_poses_topic"              default="/$(arg namespace)/$(arg app)/particle_poses"/>
  <arg name="particle_sim_mbes_topic"           default="/$(arg namespace)/$(arg app)/sim_mbes"/>
  <arg name="average_mbes_topic"                default="/$(arg namespace)/$(arg app)/avg_mbes"/>

  <arg name="rbpf_sensor_FLS" default="true"/>
  <arg name="rbpf_sensor_MBES" default="false"/>
  <arg name="survey_area_topic" default="/multi_agent/survey_area"/>
  <arg name="num_auvs" default="1"/>
  <arg name="vehicle_model" default="hugin"/>

  <!-- <arg name="color_seed" default="0"/> -->


  <!-- Rosbags from lolo -->
  <!-- <node pkg="rosbag" type="play" name="mission_player" output="log" args="-s 1050 $(arg path)/2022-06-07-12-22-19.bag"/> -->
  

  <!-- <node type="static_transform_publisher" pkg="tf" name="tf_utm_to_map" args="0 0 0 0 0 0 utm map 10"/>
  <node type="static_transform_publisher" pkg="tf" name="tf_map_to_odom" args="0 0 0 0 0 0 map $(arg odom_frame) 10"/> -->
  <!-- <include file="$(find lolo_description)/launch/lolo_description.launch" /> -->

  <!-- <param name="robot_description"
        command="$(find xacro)/xacro '$(find lolo_description)/urdf/rviz.xacro' robot_namespace:=$(arg namespace)" /> -->

  <group ns="$(arg app)">
    <group ns="$(arg namespace)">
      
      <!-- <node type="particles_launcher.py" pkg="rbpf_slam" name="particle_launcher" output="screen">
        <param name="namespace" value="$(arg namespace)"/>  
        <param name="n_beams_mbes" value="$(arg n_beams_mbes)"/>  
        <param name="results_path" value="$(arg results_path)" />
        <param name="svgp_minibatch_size" value="$(arg svgp_minibatch_size)"/>     
        <param name="num_particle_handlers" value="$(arg num_particle_handlers)"/>     
        <param name="num_particles_per_handler" value="$(eval arg('particle_count') / arg('num_particle_handlers'))"/>     
        <param name="particle_launch_file" value="$(find rbpf_slam)/launch/particle.launch"/>     
      </node> -->

      <node type="rbpf_multiagent_node" pkg="rbpf_multiagent" name="rbpf_multiagent_node" output="screen">
        <param name="particle_count"          type= "int"     value="$(arg particle_count)" />
        <param name="particle_count_neighbours" type= "int"   value="$(arg particle_count_neighbours)" />
        <!-- <rosparam param="init_covariance">[1., 1., 0.0, 0.0, 0.0, 0.1]</rosparam> -->
        <!-- <rosparam param="motion_covariance">[0.0, 0.0, 0.0, 0.0, 0.0, 0.000001]</rosparam>
        <rosparam param="resampling_noise_covariance">[3.0, 3.0, 0.0, 0.0, 0.0, 0.0]</rosparam> -->
        <rosparam param="init_covariance">[0.0, 0.0, 0.0, 0.0, 0.0, 0.0]</rosparam>
        <rosparam param="motion_covariance">[0.0, 0.0, 0.0, 0.0, 0.0, 0.0]</rosparam>
        <rosparam param="resampling_noise_covariance">[0.0, 0.0, 0.0, 0.0, 0.0, 0.0]</rosparam>
        <param name="measurement_std"                         value="10." />
        <param name="odometry_topic"                          value="$(arg odometry_topic)" />
        <param name="mbes_pings_topic"                        value="$(arg mbes_pings_topic)" />
        <param name="map_frame"                               value="$(arg map_frame)" />
        <param name="odom_frame"                              value="$(arg odom_frame)" />
        <param name="mbes_link"                               value="$(arg namespace)/mbes_link" />
        <param name="base_link"                               value="$(arg namespace)/base_link" />
        <param name="synch_topic"                             value="/$(arg app)/synch" />
        <param name="results_path"                            value="$(arg results_path)" />
        <param name="average_pose_topic"                      value="$(arg average_pose_topic)" />
        <param name="dr_pose_topic"                      value="/$(arg namespace)/dr/corrupted" />
        <param name="average_mbes_topic"                      value="$(arg average_mbes_topic)" />
        <param name="particle_poses_topic"                    value="$(arg particle_poses_topic)" />
        <param name="particle_sim_mbes_topic"                 value="$(arg particle_sim_mbes_topic)" />
        <param name="n_beams_mbes"   type= "int"              value="$(arg n_beams_mbes)" />       
        <param name="survey_finished_top" value="$(arg survey_finished_top)" />       
        <param name="pf_stats_top" value="/stats/data" />  
        <param name="manipulate_gp_server" value="$(arg manipulate_gp_server)"/>  
        <param name="minibatch_gp_server" value="$(arg gp_mb_server)"/>    
        <param name="rbpf_period" value="$(arg rbpf_period)"/> 
				<param name="rviz_period"  value="$(arg rviz_period)" />
        <param name="inducing_points_top" value="$(arg gp_ip_topic)"/>     
				<!-- <param name="path_topic"  value="/$(arg namespace)/ctrl/mission_waypoints" /> -->
				<param name="path_topic"  value="/$(arg namespace)/waypoints" />
				<param name="gp_resampling_top"  value="$(arg gp_resample_top)" />
				<param name="svgp_minibatch_size"  value="$(arg svgp_minibatch_size)" />
				<param name="particle_enable_lc"  value="/$(arg namespace)/enable_lc" />
				<param name="rbpf_saved_top"  value="/$(arg namespace)/rbpf_saved" />
				<param name="markers_top"  value="/$(arg namespace)/rbpf/particle_markers" />
				<param name="markers_left_top"  value="/$(arg namespace)/rbpf/particle_markers_left" />
				<param name="markers_right_top"  value="/$(arg namespace)/rbpf/particle_markers_right" />
        <param name="rbpf_sensor_FLS" value="$(arg rbpf_sensor_FLS)"/>
        <param name="rbpf_sensor_MBES" value="$(arg rbpf_sensor_MBES)"/>
        <param name="namespace" value="$(arg namespace)"/>
        <param name="fls_meas_topic" value="/sim/$(arg namespace)/fls_measurement" />
        <param name="survey_area_topic" value="$(arg survey_area_topic)" />
        <param name="num_auvs" value="$(arg num_auvs)"/>
        <param name="vehicle_model" value="$(arg vehicle_model)"/>


      </node>
    </group>
  </group>


  <!-- <node type="visual_tools.py" pkg="rbpf_slam" name="visual_tools" output="screen">
      <param name="particle_count" type= "int" value="$(arg particle_count)" />
      <param name="pf_stats_top" value="/stats/data" />       
      <param name="map_frame" value="$(arg map_frame)" />       
      <param name="odom_frame" value="$(arg odom_frame)" />       
      <param name="survey_finished_top" value="$(arg survey_finished_top)" />       
      <param name="result_path" value="$(arg results_path)" />
      <param name="mbes_pings_topic" value="/$(arg mode)/mbes_pings" />
      <param name="particle_sim_mbes_topic" value="$(arg particle_sim_mbes_topic)" />
   </node> -->

  <!-- Aux nodes for Lolo interface -->
  <!-- <node type="fix_odom.py" pkg="lolo_pf_interface" name="fix_odom" output="screen">
    <param name="survey_finished_top" value="$(arg survey_finished_top)" />       
    <param name="results_path" value="$(arg results_path)" />
    <param name="heading_noise" value="0.00005" />
  </node> -->

  <!-- <node type="fix_mbes_node" pkg="lolo_pf_interface" name="fix_mbes_node" output="screen"/> -->
  
  <!-- <node type="fake_synch.py" pkg="lolo_pf_interface" name="fake_synch_node" output="screen">
    <param name="survey_finished_top" value="$(arg survey_finished_top)" />       
    <param name="odometry_topic" value="$(arg odometry_topic)" />       
  </node> -->

    <!-- UW environment -->
  	<!-- <arg name="mbes_open_angle" value="1.050" />  -->
    <!-- <include file="$(find auv_model)/launch/auv_environment.launch" >
      <arg name="mode" value="$(arg mode)"/>
      <arg name="namespace" value="$(arg namespace)"/>
      <arg name="debug" default="0"/>
      <arg name="app" value="$(arg app)"/>
      <arg name="dataset" value="$(arg dataset)" />
      <arg name="n_beams_mbes" value="$(arg n_beams_mbes)"/>
      <arg name="mbes_open_angle" value="$(arg mbes_open_angle)" />
      <arg name="x" value="$(arg x)"/>
      <arg name="y" value="$(arg y)"/>
      <arg name="z" value="$(arg z)"/>
      <arg name="roll" value="$(arg roll)"/>
      <arg name="pitch" value="$(arg pitch)"/>
      <arg name="yaw" value="$(arg yaw)"/>
    </include> -->

</launch>

